index.bs

```diff
@@ -677,14 +692,515 @@ id=] must be the same as the [=window handle=].
+  IteratorObject //
+  GeneratorObject //
+  ErrorObject //
+  ProxyObject //
+  PromiseObject //
+  TypedArrayObject //
+  ArrayBufferObject //
+  NodeObject //
+  WindowProxyObject //
+}
+
+ObjectId = text;
+
+UndefinedObject = {
+  type: "undefined",
+  value: null,
+}
+
+NullObject = {
+  type: "null",
+  value: null,
+}
+
+StringObject = {
+  type: "string",
+  value: text,
+}
+
+SpecialNumber = "NaN" / "-0" / "+Infinity" / "-Infinity";
+
+NumberObject = {
+  type: "number",
+  value: number / SpecialNumber,
+}
+
+BooleanObject = {
+  type: "boolean",
+  value: bool,
+}
+
+BigIntObject = {
+  type: "bigint",
+  value: text,
+}
+
+SymbolObject = {
+  type: "symbol",
+  objectId: ObjectId,
+}
+
+ArrayObject = {
+  type: "array",
+  objectId: ObjectId,
+  value?: [RemoteObject]
+}
+
+ObjectObject = {
+  type: "object",
+  objectId: ObjectId,
+  value?: {*text: RemoteObject}
+}
+
+FunctionObject = {
+  type: "function",
+  objectId: ObjectId,
+}
+
+RegexObject = {
+  type: "regex",
+  objectId: ObjectId,
+  value: text
+}
+
+DateObject = {
+  type: "date",
+  objectId: ObjectId,
+  value: text
+}
+
+MapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+SetObject = {
+  type: "set",
+  objectId: ObjectId
+}
+
+WeakMapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+WeakSetObject = {
+  type: "set",
+  objectId: ObjectId,
+}
+
+IteratorObject = {
+  type: "iterator",
+  objectId: ObjectId,
+}
+
+GeneratorObject = {
+  type: "iterator",
+  objectId: ObjectId,
+}
+
+ErrorObject = {
+  type: "error",
+  objectId: ObjectId,
+}
+
+ProxyObject = {
+  type: "proxy",
+  objectId: ObjectId,
+}
+
+PromiseObject = {
+  type: "promise",
+  objectId: ObjectId,
+}
+
+TypedArrayObject = {
+  type: "typedarray",
+  objectId: ObjectId,
+}
+
+ArrayBufferObject = {
+  type: "arraybuffer",
+  objectId: ObjectId,
+}
+
+NodeObject = {
+  type: "node",
+  objectId: ObjectId,
+  value?: NodeValue,
+}
+
+NodeValue = {
+  nodeType: uint,
+  nodeName: text,
+  localName: text,
+  nodeValue: text,
+  childNodeCount: uint,
+  children?: [*NodeObject],
+  attributes?: {*text: text},
+}
+
+WindowProxyObject = {
+  type: "window",
+  objectId: ObjectId,
+}
+
+
+```
+
+Issue: Add WASM types?
+
+Issue: NodeValue needs more attributes
+
+Issue: Should WindowProxy get attributes in a similar style to Node?
+
+<div algorithm>
+
+To <dfn>serialize as a remote object</dfn> given an |object|, a |max depth|,
+|node details|, and a |set of known objects|:
+
+  1. Switch on the type of |object|:
+
+    <dl>
+      <dt>Undefined
+      <dd>Let |remote object| be an [=object=] matching the `UndefinedObject`
+      production in the [=local end definition=].
+
+      <dt>Null
+      <dd>Let |remote object| be an [=object=] matching the `NullObject`
+      production in the [=local end definition=].
+
+      <dt>String
+      <dd>Let |remote object| be an [=object=] matching the `StringObject`
+      production in the [=local end definition=], with the <code>value</code>
```

Comment:
Here we'll need to handle what happens if the string has lone surrogates :(

Reply:
We could spec it so that the serialization re-uses ECMAScript’s `JSON.stringify` — it was recently updated to ensure well-formed output even when lone surrogates are present. https://github.com/tc39/proposal-well-formed-stringify

Reply:
Brilliant, we should ensure that any serialization we use ends up going through this.

Reply:
It seems a little tricky to reuse directly since afaict the entrypoints all also end up quoting the string, which isn't what we want.